### ISSUE TITLE
Only allow loading a specific version/commit if it is part of the current version's log

### ISSUE THREAD

**Issue by stbrody** [ceramicnetwork/js-ceramic#614]

**Comment by stbrody** [comment]
This is actually kind of tricky to do well @oed.  The only way to load the "current" version is to wait some amount of time to hear about the tip from the pub/sub network.  Doing this the naive way risks causing loading a document at a version/commit to fail non-deterministically.  Imagine a user's js-ceramic node authors an update and submits it to an anchor service to be anchored.  The CAS tries to load the document at the given commit that's about to be anchored, but it queries against its own js-ceramic node that is not the one that authored the update.  The ceramic node that the CAS is using loads a version of the document without the new update and so errors when trying to load the version that the CAS requested since it doesn't find that version in the document's log.

To do this right we'd need a way to specify a timeout for how long you are willing to wait, and more complicated logic to detect when we've learned enough to definitively prove that the version we're requesting cannot be in the document's log so we can return with an error rather than waiting longer.

I wonder if this can wait until after release?

**Comment by oed** [comment]
Well one thing that we should do for sure, and I added a comment for it in #594, is to check at least against the local version if present.

I think loading a document at a specific version should always trigger loading the latest version as well. The non-determinism here is not great though. Note however that we still have the same problem when loading a documents latest version anyway.

**Comment by stbrody** [comment]
Okay, I think I have a way in mind that makes this work better, but on a best-effort basis.  Basically we load whatever we can get for the "current" version.  If what we load is an ancestor of the version requested, then we apply the log on top until we have the version that was requested, and don't error.  If however we call `_handleTip` with the commit being requested, and then get back a document that *doesn't* include the tip we just handled, that means that the conflict resolution mechanism must have rejected the commit we requested, so in that case we error.  Basically whenever we can definitively prove that the requested commit is invalid we error, but if we can't definitively prove that it's valid, we err on the side of succeeding
